### PR TITLE
Usbmanager: Implement parentassigngrp

### PR DESCRIPTION
**Pull request by christoph-zededa**
This PR implements usage of parentassigngrp in the model manifest.
With this new feature dependencies between IOBundles can be defined; so here for USB passthrough.

Now one edge application can use passthrough of the PCI controller and another edge application can use USB passthrough of the hardware.
Of course they cannot run and use USB passthrough at the same time, but they can use their implementation of passthrough alternating.

### CODE REVIEW

File: pkg/pillar/types/assignableadapters.go [hunk 1]
```diff
@@ -51,6 +51,13 @@ type IoBundle struct {
 	// If this is an empty string it means the IoBundle can not be assigned.
 	AssignmentGroup string
 
+	// Parent Assignment Group is there to reference the parent assignment group in order to make the device
+	// dependent on a different device.
+	// Currently the concrete reason to do this is to make a usb device dependent on the PCI address the USB
+	// controller is using to provent passthrough of the USB controller in one application while trying to passthrough
```
Comment (rouming) [inline pkg/pillar/types/assignableadapters.go:57]: s/provent/prevent/ ?

File: pkg/pillar/cmd/usbmanager/ioBundleTree.go [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+
+	ibe.parent = iobt.root
+}
+
+func (iobt *ioBundleTree) addIOBundle(ioBundle *types.IoBundle) {
+	dependents := iobt.groupDependendents(ioBundle.AssignmentGroup)
+	for _, dependee := range dependents {
+		if dependee == ioBundle.ParentAssignmentGroup {
+			log.Warnf("detected parentassigngrp circular dependency, not adding ioBundle %s", ioBundle.Phylabel)
+			return
```
Comment (christoph-zededa) [inline pkg/pillar/cmd/usbmanager/ioBundleTree.go:157]: return error

Comment (christoph-zededa) [reply pkg/pillar/cmd/usbmanager/ioBundleTree.go:157]: +tell cloud

File: pkg/pillar/cmd/usbmanager/ioBundleTree.go [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+	}
+
+	if ibe.parent == nil && ibe.assignmentGroup != "" {
+		ibe.parent = parentIbe
+	}
+
+	ibe.ioBundlesMap[ioBundle.Phylabel] = ioBundle
+
+	if ibe.parent != nil && ibe.parent.assignmentGroup != ioBundle.ParentAssignmentGroup {
+		if len(ibe.ioBundlesMap) == 1 {
```
Comment (christoph-zededa) [inline pkg/pillar/cmd/usbmanager/ioBundleTree.go:204]: == 0

File: pkg/pillar/cmd/usbmanager/ioBundleTree.go [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+			// add ioBundle with Parent X
+			// ioBundleElem with ioBundle and parent ioBundleElem X is added
+			// remove this ioBundle
+			// ioBundleElem still exists and the parent still points to ioBundleElem X
+			// add ioBundle with Parent Y
+			// now we set the parent of the ioBundleElemn to the Y ioBundleElem
+
+			ibe.parent = parentIbe
+		} else {
+			panic("this should not happen, seems two ioBundles with same assignment group have two different parent assignment groups")
```
Comment (christoph-zededa) [inline pkg/pillar/cmd/usbmanager/ioBundleTree.go:217]: replace with warning

File: pkg/pillar/cmd/usbmanager/usbcontroller.go [hunk 5]
```diff
@@ -148,41 +262,66 @@ func (uc *usbmanagerController) addVirtualmachine(vm virtualmachine) {
 
 	// add rules
 	for _, phyLabel := range vm.adapters {
-		ioBundle := uc.ioBundles[phyLabel]
+		ioBundle := uc.iobt.ioBundle(phyLabel)
 		if ioBundle == nil {
 			continue
 		}
+		neighborIOBundlesElem := uc.iobt.elementsByAssignmentGroup[ioBundle.AssignmentGroup]
```
Comment (christoph-zededa) [inline pkg/pillar/cmd/usbmanager/usbcontroller.go:269]: siblings

File: pkg/pillar/cmd/usbmanager/usbcontroller_test.go [hunk 6]
```diff
@@ -254,169 +254,6 @@ func TestConnectUSBDevicesToQemu(t *testing.T) {
 	}
 }
 
-func TestIOBundleEmpty(t *testing.T) {
```
Comment (christoph-zededa) [inline pkg/pillar/cmd/usbmanager/usbcontroller_test.go:255]: create extra commit moving the tests

File: pkg/pillar/cmd/domainmgr/domainmgr.go [hunk 7]
```diff
@@ -1334,7 +1334,6 @@ func handleCreate(ctx *domainContext, key string, config *types.DomainConfig) {
 }
 
 // doAssignAdaptersToDomain assigns IO adapters to the newly created domain.
-// Note that the adapters are already reserved for the domain using reserveAdapters (UsedByUUID is set).
```
Comment (eriknordmark) [inline pkg/pillar/cmd/domainmgr/domainmgr.go:1337]: Did you change reserveAdapters to not sed UsedByUUID?

**Review by eriknordmark — APPROVED** [review]
Run eden

File: pkg/pillar/cmd/domainmgr/domainmgr.go (hunk 7 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/domainmgr/domainmgr.go:1337]: Looking at this some more you still set that, but the hyper.pcireserve is deferred. So it makes more sense to ammend the comment than to delete it.

**Review by eriknordmark — COMMENTED** [review]
I'm far from sure the many commits makes sense as a standalone commit.
My personal thoughts on this is that if one needs a some new functionality in package X or file Y, a separate commit makes sense for introducing this (with a comment it will be used later to do XYZ), followed by a commit which uses this new functionality.

Thus one can pick any point in the sequence of commits and build a working system.
I don't know if my personal views match others.

But here we have a case of e.g., moving the call from hyper,pcireserve from one place to another which is split in two commits (removing it from the old and adding it to the new), thus the intermediate state might build in this case, but will not work AFAIU.

What is a resonable set of commits for this? There are a set of improvements in usbmanager, changes in domainmgr to use them, docs, etc. Thoughts?

**Comment by rouming** [comment]
> I don't know if my personal views match others.

Absolute match.

Ideally every commit in the patchset should not break anything (including a build). This is especially important when someone in the future bisects in order to find an introduced bug and curses when in the middle of bisect procedure the build is broken. Applying this principle to any changes introduced in internal calls (new functions, new API, etc) can be tricky, but the rule of thumb is usually is the following:
1. introduce new library / set of calls / API (series of commits)
2. switch to new library / set of calls / API (1 commit)
3. remove old library / set of calls / API (1 commit or series of commits)

**Comment by christoph-zededa** [comment]
> I'm far from sure the many commits makes sense as a standalone commit. My personal thoughts on this is that if one needs a some new functionality in package X or file Y, a separate commit makes sense for introducing this (with a comment it will be used later to do XYZ), followed by a commit which uses this new functionality.

The good thing is, combining commits is always easier than splitting it up :-)

Some months ago I was bisecting an issue and I found a commit that was just a fix in response to a PR comment. Before that I thought it would always be good to fixup such commits, but in this specific case it was advantageous to not have.

> 
> Thus one can pick any point in the sequence of commits and build a working system. I don't know if my personal views match others.

I think it should be buildable and the tests (go test) should succeed. 
I did a `git rebase --interactive --exec "make test" origin/master` in the past, but not with the current code in this PR (also because there is a flaky test) and I am pretty sure https://github.com/lf-edge/eve/pull/3755/commits/10bb57c63a4a89cca0a9b27158934eb7e38936a2 will not succeed.

IMHO if we want that every commit is buildable then our github actions should not only build the HEAD of the PR but every commit.

> 
> But here we have a case of e.g., moving the call from hyper,pcireserve from one place to another which is split in two commits (removing it from the old and adding it to the new), thus the intermediate state might build in this case, but will not work AFAIU.

You mean https://github.com/lf-edge/eve/pull/3755/commits/35bde1b51a7911573842680c5ab624ee8283d9ce ?
But I don't see where it is added again. The issue is, that domainmgr was reserving the device twice:
* at https://github.com/lf-edge/eve/blob/master/pkg/pillar/cmd/domainmgr/domainmgr.go#L1390
* at https://github.com/lf-edge/eve/blob/master/pkg/pillar/cmd/domainmgr/domainmgr.go#L3108

So I only need to remove it once.

**Comment by eriknordmark** [comment]
> > But here we have a case of e.g., moving the call from hyper,pcireserve from one place to another which is split in two commits (removing it from the old and adding it to the new), thus the intermediate state might build in this case, but will not work AFAIU.
> 
> You mean [35bde1b](https://github.com/lf-edge/eve/commit/35bde1b51a7911573842680c5ab624ee8283d9ce) ? But I don't see where it is added again. 

Yes. And you  deleted it from another function in https://github.com/lf-edge/eve/pull/3755/commits/0fcd9336d7db65c09910fe5cb588ade98d1f7eb7
The combination of those two is a move, and just the delete or add doesn't make any sense; the move makes sense.

File: pkg/pillar/cmd/domainmgr/domainmgr.go (hunk 7 above)
Comment (christoph-zededa) [reply pkg/pillar/cmd/domainmgr/domainmgr.go:1337]: I changed the comment to:
```
// doAssignAdaptersToDomain assigns IO adapters to the newly created domain.                                                                                      
// The adapters are reserved here for the domain                                                                                                                  
// UsedByUUID is already set in reserveAdapters            
```

**Comment by christoph-zededa** [comment]
> > > But here we have a case of e.g., moving the call from hyper,pcireserve from one place to another which is split in two commits (removing it from the old and adding it to the new), thus the intermediate state might build in this case, but will not work AFAIU.
> > 
> > 
> > You mean [35bde1b](https://github.com/lf-edge/eve/commit/35bde1b51a7911573842680c5ab624ee8283d9ce) ? But I don't see where it is added again.
> 
> Yes. And you deleted it from another function in [0fcd933](https://github.com/lf-edge/eve/commit/0fcd9336d7db65c09910fe5cb588ade98d1f7eb7) The combination of those two is a move, and just the delete or add doesn't make any sense; the move makes sense.

I combined it into: https://github.com/lf-edge/eve/pull/3755/commits/4049c2abfec2baafecf995650c8de5a977874a6d

**Review by rouming — APPROVED** [review]

File: pkg/pillar/cmd/domainmgr/domainmgr.go (hunk 7 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/domainmgr/domainmgr.go:1337]: Thanks

**Comment by eriknordmark** [comment]
> I combined it into: [4049c2a](https://github.com/lf-edge/eve/commit/4049c2abfec2baafecf995650c8de5a977874a6d)

Sorry for not being clear, but that was merely the most obvious example that these many separate commits will not make any sense to anybody reading the history. Maybe there are 3-6 separate items in here, but not 19.

**Comment by christoph-zededa** [comment]
> > I combined it into: [4049c2a](https://github.com/lf-edge/eve/commit/4049c2abfec2baafecf995650c8de5a977874a6d)
> 
> Sorry for not being clear, but that was merely the most obvious example that these many separate commits will not make any sense to anybody reading the history. Maybe there are 3-6 separate items in here, but not 19.

I now combined it into 6 commits.

**Review by eriknordmark — APPROVED** [review]
Re-run eden

**Pull request merged**